coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala

```diff
@@ -80,11 +84,60 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
     ) { // Query was part of routing
       localPartitionPlanner.materialize(logicalPlan, qContext)
     } else logicalPlan match {
-      case mqp: MetadataQueryPlan             => materializeMetadataQueryPlan(mqp, qContext).plans.head
-      case lp: TsCardinalities                => materializeTsCardinalities(lp, qContext).plans.head
-      case _                                  => walkLogicalPlanTree(logicalPlan, qContext).plans.head
+      case mqp: MetadataQueryPlan  => materializeMetadataQueryPlan(mqp, qContext).plans.head
+      case lp: TsCardinalities     => materializeTsCardinalities(lp, qContext).plans.head
+      case lp                       => {
+        val queryParams = qContext.origQueryParams.asInstanceOf[PromQlQueryParams]
+        // Get all the valid time-ranges to query.
+        //   For example, cross-partition lookback is unsupported. These time-ranges would
+        //   indicate queryable ranges where all involved lookbacks (range-selector windows,
+        //   stale-data lookbacks, etc) do not cross partitions.
+        val timeRanges = {
+          val invalidRanges = getInvalidRanges(lp, queryParams)
+          val mergedInvalidRanges = mergeAndSortRanges(invalidRanges)
+          val totalTimeRange = TimeRange(1000 * queryParams.startSecs, 1000 * queryParams.endSecs)
+          invertRanges(mergedInvalidRanges, totalTimeRange)
+        }
+        // Generate (logicalPlan, qContext) pairs for each of the above time-ranges, then materialize.
+        // If only one valid time range exists, it spans only a single second, and the range matches
+        //   the queryParams, we assume this is intentionally an instant query and do not update
+        //   the plan via "copyLogicalPlanWithUpdatedTimeRange".
+        // copyLogicalPlanWithUpdatedTimeRange would set all nested plan ranges to exactly one second,
+        //   which would break the instant query.
```

Comment:
TODO: this is not true, but it does cause issues with `TopLevelSubquery` (and possibly others).